spec/models/job_preferences/job_scope_spec.rb

```diff
@@ -0,0 +1,205 @@
+    create(:vacancy,
+           :published,
+           job_role: "senior_leader",
+           phases: %w[primary],
+           key_stages: %w[ks2],
+           subjects: [],
+           working_patterns: %w[full_time])
+  end
+
+  let!(:vacancy4) do
+    create(:vacancy,
+           :published,
+           job_role: "middle_leader",
+           phases: %w[secondary],
+           key_stages: %w[ks1 ks2],
+           subjects: %w[Mathematics French Spanish],
+           working_patterns: %w[part_time full_time])
+  end
+
+  let(:job_preferences) do
+    build_stubbed(:job_preferences,
+                  roles: %w[teacher teaching_assistant head_of_year headteacher],
+                  phases: %w[primary secondary],
+                  working_patterns: %w[part_time full_time],
+                  key_stages: %w[ks1 ks2])
+  end
+
+  subject(:job_scope) { described_class.new(original_scope, job_preferences) }
+
+  describe "roles scope" do
+    let(:jobseeker_roles) { %w[teacher teaching_assistant] }
+
+    before { job_preferences.roles = jobseeker_roles }
+
+    it "returns vacancies with a role matching any of the jobseeker roles" do
+      expect(job_scope.call).to contain_exactly(vacancy1, vacancy2)
+    end
+
+    %w[head_of_year head_of_department].each do |granular_role|
+      context "with a jobseeker '#{granular_role}' role split from 'middle_leader' role" do
+        let(:jobseeker_roles) { ["teacher", granular_role] }
+
+        it "returns vacancies with 'middle_leader' role" do
+          expect(job_scope.call).to contain_exactly(vacancy1, vacancy4)
+        end
+      end
+    end
+
+    %w[headteacher headteacher_deputy headteacher_assistant].each do |granular_role|
+      context "with a jobseeker '#{granular_role}' role split from 'senior_leader' role" do
+        let(:jobseeker_roles) { ["teacher", granular_role] }
+
+        it "returns vacancies with 'senior_leader' role" do
+          expect(job_scope.call).to contain_exactly(vacancy1, vacancy3)
+        end
+      end
+    end
+
+    context "with jobseeker roles that don't match any vacancy roles" do
+      let(:jobseeker_roles) { %w[sendco] }
+
+      it "returns no vacancies" do
+        expect(job_scope.call).to be_empty
+      end
+    end
+
+    context "with jobseeker without roles" do
+      let(:jobseeker_roles) { [] }
+
+      it "returns no vacancies" do
+        expect(job_scope.call).to be_empty
+      end
+    end
+  end
+
+  describe "phases scope" do
+    before { job_preferences.phases = jobseeker_phases }
+
+    context "with all jobseeker phases matching vacancy phases" do
+      let(:jobseeker_phases) { %w[primary secondary] }
+
+      it "returns any vacancies with a phase matching the jobseeker phases" do
+        expect(job_scope.call).to contain_exactly(vacancy1, vacancy2, vacancy3, vacancy4)
+      end
+    end
+
+    context "with some of the jobseeker phases matching vacancy phases" do
+      let(:jobseeker_phases) { %w[primary middle] }
+
```

Comment:
I know this isn't related to this PR but shouldn't this show `vacancy1` and `vacancy2` as well as they both include `primary`?

Reply:
Thanks for trying to understand every assertion here :clap: .

This is why I decided/needed to test the whole `JobScope` behaviour. The scoping logic is mainly done through `PostgreSQL` array methods and is not immediately obvious.

For the phases scoping, the `JobScope` applies this query:
```
.where("phases <@ ARRAY[?]::int[]", Vacancy.phases.values_at(*job_preferences.phases))
```

From [PostgreSQL Array functions page](https://www.postgresql.org/docs/current/functions-array.html)

```
anyarray <@ anyarray → boolean

Is the first array contained by the second?

ARRAY[2,2,7] <@ ARRAY[1,7,4,2,6] → 
```

It filters out any vacancy with multiple phases if all the vacancy phases are not selected in the jobseeker preferences. 
As `vacancy1` and `vacancy2` have 2 phases: `["primary", "secondary"]` and the jobseeker preferences are `["primary", "middle"]`, vacancies are excluded from the results due to jobseeker preferences not including `"secondary"`.

The behaviour you expect corresponds to the scopes using `&&` operator instead of `<@`. EG: The `subjects` scoping is done that way.